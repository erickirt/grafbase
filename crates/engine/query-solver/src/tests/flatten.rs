use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/requires-with-argument/a")
    B @join__graph(name: "b", url: "http://localhost:4200/requires-with-argument/b")
    C @join__graph(name: "c", url: "http://localhost:4200/requires-with-argument/c")
    D @join__graph(name: "d", url: "http://localhost:4200/requires-with-argument/d")
}

type Author implements Node
    @join__type(graph: D)
{
    id: ID!
    name: String
}

type Comment implements Node
    @join__type(graph: C, key: "id")
    @join__type(graph: D, key: "id")
{
    id: ID!
    authorId: ID @join__field(graph: C)
    body: String! @join__field(graph: C)
    date: String @join__field(graph: D)
}

type Post implements Node
    @join__type(graph: C, key: "id")
    @join__type(graph: D, key: "id")
{
    id: ID!
    author: Author @join__field(graph: D, requires: "comments(limit: 3) { authorId }")
    comments(limit: Int!): [Comment] @join__field(graph: D)
}

type Product
    @join__type(graph: A, key: "upc")
    @join__type(graph: B, key: "upc")
{
    upc: String!
    weight: Int @join__field(graph: B)
    price(currency: String!): Int @join__field(graph: B)
    shippingEstimate: Int @join__field(graph: A, requires: "price(currency: \"USD\") weight")
    name: String @join__field(graph: B)
}

interface Node {
    id: ID!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
    @join__type(graph: C)
    @join__type(graph: D)
{
    node: Node @join__field(graph: D)
    products: [Product] @join__field(graph: B)
    feed: [Post] @join__field(graph: C)
}
"#;

#[test]
fn comments_fields_should_be_flattened() {
    assert_solving_snapshots!(
        "comments_fields_should_be_flattened",
        SCHEMA,
        r#"
        query ($limit: Int = 1) {
          feed {
            author {
              id
            }
            ...Foo
            ...Bar
          }
        }

        fragment Foo on Post {
          comments(limit: $limit) {
            id
          }
        }

        fragment Bar on Post {
          comments(limit: $limit) {
            id
          }
        }
        "#
    );
}

#[test]
fn flattening_cannot_merge_fields_with_different_type_conditions() {
    assert_solving_snapshots!(
        "flattening_cannot_merge_fields_with_different_type_conditions",
        SCHEMA,
        r#"
        query {
          node {
            ... on Post {
              ... on Node {
                id
              }
            }
            ... on Comment {
              ... on Node {
                id
              }
            }
          }
        }
        "#
    );
}
