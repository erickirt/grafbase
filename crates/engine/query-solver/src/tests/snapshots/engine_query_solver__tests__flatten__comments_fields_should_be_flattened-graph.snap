---
source: crates/engine/query-solver/src/tests/flatten.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.feed\" ]\n    2 [ label = \"Post.author\" ]\n    3 [ label = \"Post.comments\" ]\n    4 [ label = \"Author.id\" ]\n    5 [ label = \"Comment.id\" ]\n    6 [ label = \"Root#c\", shape=parallelogram, color=dodgerblue ]\n    7 [ label = \"feed#c\", shape=box, color=dodgerblue ]\n    8 [ label = \"FedEntity#d\", shape=parallelogram, color=dodgerblue ]\n    9 [ label = \"author#d\", shape=box, color=dodgerblue ]\n    10 [ label = \"id#d\", shape=box, color=dodgerblue ]\n    11 [ label = \"comments#d\", shape=box, color=dodgerblue ]\n    12 [ label = \"id#d\", shape=box, color=dodgerblue ]\n    13 [ label = \"*Post.comments\" ]\n    14 [ label = \"comments#d\", shape=box, color=dodgerblue ]\n    15 [ label = \"*Comment.authorId\" ]\n    16 [ label = \"FedEntity#c\", shape=parallelogram, color=dodgerblue ]\n    17 [ label = \"authorId#c\", shape=box, color=dodgerblue ]\n    18 [ label = \"*Comment.id\" ]\n    19 [ label = \"id#d\", shape=box, color=dodgerblue ]\n    20 [ label = \"*Post.id\" ]\n    21 [ label = \"id#c\", shape=box, color=dodgerblue ]\n    0 -> 1 [  ]\n    1 -> 2 [  ]\n    1 -> 3 [  ]\n    2 -> 4 [  ]\n    3 -> 5 [  ]\n    0 -> 6 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 6 [ style=dashed,arrowhead=none ]\n    6 -> 7 [ color=royalblue,fontcolor=royalblue ]\n    7 -> 1 [ color=violet,arrowhead=none ]\n    7 -> 8 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 8 [ style=dashed,arrowhead=none ]\n    8 -> 9 [ color=royalblue,fontcolor=royalblue ]\n    9 -> 2 [ color=violet,arrowhead=none ]\n    9 -> 10 [ color=royalblue,fontcolor=royalblue ]\n    10 -> 4 [ color=violet,arrowhead=none ]\n    8 -> 11 [ color=royalblue,fontcolor=royalblue ]\n    11 -> 3 [ color=violet,arrowhead=none ]\n    11 -> 12 [ color=royalblue,fontcolor=royalblue ]\n    12 -> 5 [ color=violet,arrowhead=none ]\n    1 -> 13 [  ]\n    9 -> 13 [ color=orangered,arrowhead=inv ]\n    8 -> 14 [ color=royalblue,fontcolor=royalblue ]\n    14 -> 13 [ color=violet,arrowhead=none ]\n    13 -> 15 [  ]\n    9 -> 15 [ color=orangered,arrowhead=inv ]\n    14 -> 16 [ color=royalblue,fontcolor=royalblue ]\n    13 -> 16 [ style=dashed,arrowhead=none ]\n    16 -> 17 [ color=royalblue,fontcolor=royalblue ]\n    17 -> 15 [ color=violet,arrowhead=none ]\n    13 -> 18 [  ]\n    16 -> 18 [ color=orangered,arrowhead=inv ]\n    14 -> 19 [ color=royalblue,fontcolor=royalblue ]\n    19 -> 18 [ color=violet,arrowhead=none ]\n    1 -> 20 [  ]\n    8 -> 20 [ color=orangered,arrowhead=inv ]\n    7 -> 21 [ color=royalblue,fontcolor=royalblue ]\n    21 -> 20 [ color=violet,arrowhead=none ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.feed]
    2 [ Post.author]
    3 [ Post.comments]
    4 [ Author.id]
    5 [ Comment.id]
    6 [ Root#c]
    7 [ feed#c]
    8 [ FedEntity#d]
    9 [ author#d]
    10 [ id#d]
    11 [ comments#d]
    12 [ id#d]
    13 [ *Post.comments]
    14 [ comments#d]
    15 [ *Comment.authorId]
    16 [ FedEntity#c]
    17 [ authorId#c]
    18 [ *Comment.id]
    19 [ id#d]
    20 [ *Post.id]
    21 [ id#c]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    1 -> 3 [ label = "Field" ]
    2 -> 4 [ label = "Field" ]
    3 -> 5 [ label = "Field" ]
    0 -> 6 [ label = "CreateChildResolver" ]
    0 -> 6 [ label = "HasChildResolver" ]
    6 -> 7 [ label = "CanProvide" ]
    7 -> 1 [ label = "Provides" ]
    7 -> 8 [ label = "CreateChildResolver" ]
    1 -> 8 [ label = "HasChildResolver" ]
    8 -> 9 [ label = "CanProvide" ]
    9 -> 2 [ label = "Provides" ]
    9 -> 10 [ label = "CanProvide" ]
    10 -> 4 [ label = "Provides" ]
    8 -> 11 [ label = "CanProvide" ]
    11 -> 3 [ label = "Provides" ]
    11 -> 12 [ label = "CanProvide" ]
    12 -> 5 [ label = "Provides" ]
    1 -> 13 [ label = "Field" ]
    9 -> 13 [ label = "Requires" ]
    8 -> 14 [ label = "CanProvide" ]
    14 -> 13 [ label = "Provides" ]
    13 -> 15 [ label = "Field" ]
    9 -> 15 [ label = "Requires" ]
    14 -> 16 [ label = "CreateChildResolver" ]
    13 -> 16 [ label = "HasChildResolver" ]
    16 -> 17 [ label = "CanProvide" ]
    17 -> 15 [ label = "Provides" ]
    13 -> 18 [ label = "Field" ]
    16 -> 18 [ label = "Requires" ]
    14 -> 19 [ label = "CanProvide" ]
    19 -> 18 [ label = "Provides" ]
    1 -> 20 [ label = "Field" ]
    8 -> 20 [ label = "Requires" ]
    7 -> 21 [ label = "CanProvide" ]
    21 -> 20 [ label = "Provides" ]
}
